### PR TITLE
feat(vm): support resize pvc

### PR DESCRIPTION
**Pull request by FrankYang0529**
**Feature:**
Support resizing PVC via VM.

**Related Issue:**
https://github.com/harvester/harvester/issues/2810

**Test plan:**

### Setup

1. Create a harvester cluster.
2. Upload a VM image.

### Case 1: Resizing a volume via Volumes page doesn't break the VM.

1. Create a VM with a 10G disk.
2. Stop the VM.
3. In the Volumes page, resize the disk to 20G.
4. After the disk is resized, start the VM. We should not have any error message in `harvester/harvester` deployment.
5. Remove the VM and the disk.

### Case 2: Can't resize volumes on a running VM.

1. Create a VM with a 10G disk.
2. Update storage size to 20G in `harvesterhci.io/volumeClaimTemplates` annotation of the VM. We should get an error message like `please stop the VM before resizing volumes`.
3. Remove the VM and the disk.

### Case 3: Can't resize volumes to a smaller size.

1. Create a VM with a 10G disk.
2. Update storage size to 5G in `harvesterhci.io/volumeClaimTemplates` annotation of the VM. We should get an error message like `PVC requests storage can't be less than previous value`.
3. Remove the VM and the disk.

### Case 4: We can resize volumes through VM annotation.

1. Create a VM with a 10G disk.
2. Stop the VM.
3. Update storage size to 20G in `harvesterhci.io/volumeClaimTemplates` annotation of the VM. We should not get any error in this step.
4. After the disk is resized, start the VM. VM can run successfully.

### CODE REVIEW

**Comment by FrankYang0529** [comment]
The CodeFactor error is from complex method `SetOwnerOfPVCs`. I don't touch the function in this PR, so please ignore it. Thanks.

File: pkg/controller/master/virtualmachine/vm_controller.go [hunk 1]
```diff
@@ -51,15 +51,35 @@ func (h *VMController) createPVCsFromAnnotation(_ string, vm *kubevirtv1.Virtual
...
 				return nil, err
 			}
+			continue
 		} else if err != nil {
 			return nil, err
 		}
+
+		// Users also can resize volumes through Volumes page. In that case, we can't track the update in VM annotation.
+		// If storage request in the VM annotation is smaller than the real PVC, we skip it.
+		if pvcAnno.Spec.Resources.Requests.Storage().Cmp(*pvc.Spec.Resources.Requests.Storage()) == -1 {
```
Comment (futuretea) [inline pkg/controller/master/virtualmachine/vm_controller.go]: should be <= 0?

File: pkg/webhook/resources/virtualmachine/validator.go [hunk 2]
```diff
@@ -138,6 +140,56 @@ func (v *vmValidator) checkVolumeClaimTemplatesAnnotation(vm *kubevirtv1.Virtual
...
+	if err := json.Unmarshal([]byte(oldVM.Annotations[util.AnnotationVolumeClaimTemplates]), &oldPvcs); err != nil {
+		return werror.NewInvalidError(fmt.Sprintf("failed to unmarshal %s", oldVM.Annotations[util.AnnotationVolumeClaimTemplates]), fmt.Sprintf("metadata.annotations.%s", util.AnnotationVolumeClaimTemplates))
+	}
+	if err := json.Unmarshal([]byte(newVM.Annotations[util.AnnotationVolumeClaimTemplates]), &newPvcs); err != nil {
+		return werror.NewInvalidError(fmt.Sprintf("failed to unmarshal %s", newVM.Annotations[util.AnnotationVolumeClaimTemplates]), fmt.Sprintf("metadata.annotations.%s", util.AnnotationVolumeClaimTemplates))
+	}
+
+	oldPvcMap, newPvcMap := map[string]*corev1.PersistentVolumeClaim{}, map[string]*corev1.PersistentVolumeClaim{}
+	for _, pvc := range oldPvcs {
+		oldPvcMap[pvc.Name] = pvc
```
Comment (futuretea) [inline pkg/webhook/resources/virtualmachine/validator.go:158]: maybe need to use Namespace/Name as the map key

Comment (FrankYang0529) [reply pkg/webhook/resources/virtualmachine/validator.go:158]: A pod can't use PVC in different namespaces, so I think using PVC name is ok here.

Ref: https://discuss.kubernetes.io/t/accessing-persistent-volumes-in-pods-across-differernt-namespaces/19703

File: pkg/controller/master/virtualmachine/vm_controller.go (hunk 1 above)
Comment (FrankYang0529) [reply pkg/controller/master/virtualmachine/vm_controller.go]: Updated it. Thank you.

**Review by futuretea — APPROVED** [review]
LGTM. Thanks! tested follow the `Test Plan`

**Review by bk201 — APPROVED** [review]

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

**Comment by FrankYang0529** [comment]
@Mergifyio backport v1.1

**Pull request merged**